### PR TITLE
[SERVICE-329] API docs (Other)

### PR DESCRIPTION
**Pull request by ryanpetersonOF**
**Do not merge** until everything is settled.

This cleans up the API docs overall.  Removes the `Types` module with former content being moved into its respective module (tabbing, workspaces, etc).  Some client interfaces have been renamed or removed to declutter the docs, specifically the `someEventPayload` and `someEventEvent` duplication. 

Breaking changes:
* No longer have a `Types.ts` to import from.  Imports will be found in respective modules.

Other changes:
* No longer using window.addEventListener.  We have swapped this out to use EventEmitter instead.
* Events have been updated!  They now include a type, and are no longer CustomEvents but rather straight payloads.

Recommend pulling this down and running `npm run docs` to have a look at changes.

### CODE REVIEW

**Comment by ryanpetersonOF** [comment]
@bryangaleOF 
The Event naming is because these are "Events" coming from the service.  We can keep them payload naming, but I think it's confusing when you look at the docs as these interfaces are all listed out.  Would EventNamePayload make it clear these are for an event? 

We use `CustomEvent` because we are using the native `window.addEventListener` which will only take `Event` or `CustomEvent`.  Since we have a custom event details, we need to use `CustomEvent`.  There are two cases where we dont have custom event details and they use `Event`.

The type string is still there except now it will come in from `addEventListener(type, listener)`.  There was no need to set it the way we were doing before.

**Comment by bryangaleOF** [comment]
My previous comment got slightly manged - 'CustomEvent' was meant to be `CustomEvent<FooEvent>`. Updated.

**Comment by tomrobertsOF** [comment]
Gonna wait until PR#277 is merged in to avoid doubling up reviews.

**Comment by bryangaleOF** [comment]
> @bryangaleOF
> The Event naming is because these are "Events" coming from the service. We can keep them payload naming, but I think it's confusing when you look at the docs as these interfaces are all listed out. Would EventNamePayload make it clear these are for an event?
> 

For me, EventNamePayload would make it clear these aren't inheriting `Event`. The naming implies something about the inheritance hierarchy and I think it'd come as a surprise that Foo1Event, Foo2Event etc. don't share a common `Event` base class. But I acknowledge what's going to be clear in the docs is another concern to weigh up here.

> The type string is still there except now it will come in from `addEventListener(type, listener)`. There was no need to set it the way we were doing before.
But the `type` field inherited from `Event` on the `CustomEvent` object will remain unset if I've understood correctly. My worry is that catching someone out down the line. Like, is there an implicit contract when you have an `Event` object that you can rely on the `type` field being set correctly?

**Comment by ryanpetersonOF** [comment]
@bryangaleOF 
Your understanding is incorrect :).  The type is set by `addEventListener`, so the event type will always match the type you are listening for.

```
window.addEventListener("myCoolEvent", (e) => {
    console.log(e.type === "myCoolEvent");
});

window.dispatchEvent(new Event("myCoolEvent"));
```

In connection.ts we are creating the events.

Lets discuss naming further in the real life.

**Comment by bryangaleOF** [comment]
Just realized I missed that CustomEvent takes the type string in its constructor, which then (I assume) sets `this.type = foo`. That's where my confusion was coming from!

File: src/client/main.ts [hunk 1]
```diff
@@ -1,16 +1,51 @@
...
+     */
+    uuid: string;
+
+    /**
+     * Window identifier
+     */
+    name: string;
+}
+
+// To be updated/moved once config story is fully merged. SERVICE-306
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/main.ts]: Fixed

File: src/client/snapanddock.ts [hunk 2]
```diff
@@ -1,31 +1,18 @@
...
-}
 
 /**
  * @hidden
  */
 export interface EventMap {
-    'window-docked': WindowDockedEvent;
-    'window-undocked': WindowUndockedEvent;
+    'window-docked': Event;
+    'window-undocked': Event;
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/snapanddock.ts]: Fixed per our discussion.  Sorry, not types :(

File: src/client/connection.ts [hunk 3]
```diff
@@ -42,22 +43,31 @@ export const channelPromise: Promise<ChannelClient> = typeof fin === 'undefined'
...
+        channel.register('tab-group-minimized', (payload: TabGroupMinimizedEvent) => {
+            window.dispatchEvent(new CustomEvent<TabGroupMinimizedEvent>('tab-group-minimized', {detail: payload}));
+        });
+        channel.register('tab-group-maximized', (payload: TabGroupMaximizedEvent) => {
+            window.dispatchEvent(new CustomEvent<TabGroupMaximizedEvent>('tab-group-maximized', {detail: payload}));
+        });
 
         channel.register('workspace-generated', (payload: Workspace) => {
-            window.dispatchEvent(new CustomEvent<Workspace>('workspace-generated', {detail: payload}));
+            window.dispatchEvent(new CustomEvent<WorkspaceGeneratedEvent>('workspace-generated', {detail: {workspace: payload}}));
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/connection.ts]: Fixed with event emitter stuff

File: src/client/snapanddock.ts [hunk 4]
```diff
@@ -70,9 +65,26 @@ export async function addEventListener(eventType: 'window-docked', listener: (ev
...
+ */
+export interface EventMap {
+    'window-docked': Event;
+    'window-undocked': Event;
+}
+
+/**
+ * @type window-docked
+ */
+export async function addEventListener(eventType: 'window-docked', listener: (event: Event) => void): Promise<void>;
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/snapanddock.ts]: Done

File: src/client/main.ts [hunk 5]
```diff
@@ -29,15 +30,38 @@ export interface WindowIdentity {
...
+    /**
+     * Defines the regex pattern.
+     *
+     * Do not wrap the expression in `/` characters - specify the pattern string as you would when passing to the RegExp constructor.
+     */
     expression: string;
+
+    /**
+     * Any additional flags that form part of this expression. Supported same [values
+     * ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Parameters)as the native JS `RegExp` class.
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/main.ts]: Yep still works

File: src/client/snapanddock.ts [hunk 6]
```diff
@@ -40,12 +34,13 @@ export interface EventMap {
...
  * group. This may change in future versions of the service.
  *
- * @type window-docked
  * @event
  */
-export async function addEventListener(eventType: 'window-docked', listener: (event: Event) => void): Promise<void>;
+export interface WindowDockedEvent {}
 
 /**
+ * Details of the {@link addEventListener|'window-undocked'} event.
```
[the remark this answers is not in the document]
Comment (ryanpetersonOF) [reply src/client/snapanddock.ts]: Cleaned up

**Pull request merged**